### PR TITLE
[collector] adding basic collector and fitting pieces for collection

### PR DESCRIPTION
**Pull request by truthbk**

### CODE REVIEW

File: collector/collector.py [hunk 1]
```diff
@@ -3,8 +3,104 @@
...
+            for check_name, configs in check_configs.iteritems().iteritems():
+                check_class = self._check_classes.get(check_name)
+                if check_class:
+                    for config in configs:
+                        init_config = config.get('init_config', {})
+                        instances = config.get('instances')  # should be single instance
+                        for instance in instances:
+                            signature = (init_config, instance)
+                            if signature in self._check_instance_signatures:
+                                continue
```
Comment (hush-hush) [inline collector/collector.py]: We should print an error message: it could be confusing to have only one instance running when multiple are configured

File: collector/collector.py [hunk 2]
```diff
@@ -3,8 +3,104 @@
...
+                                        instance, check_name)
+
+    def run_checks(self):
+        for name, checks in self._check_instances.iteritems():
+            self.log.debug('running check %s...', name)
+            for check in checks:
+                try:
+                    check.run(check.instance)
+                except Exception:
+                    log.exception("error for instance: %s", str(check.instance))
```
Comment (hush-hush) [inline collector/collector.py]: Do we want to send a service_check for every instance ran ? We could do it in another PR

File: agent.py [hunk 3]
```diff
@@ -22,24 +27,52 @@ def init_agent():
...
-    logging.info("Starting the agent, hostname: %s", get_hostname())
+    hostname = get_hostname()
+
+    logging.info("Starting the agent, hostname: %s", hostname)
 
     # init Forwarder
     logging.info("Starting the Forwarder")
-    f = Forwarder(config.get("api_key"), config.get("dd_url"))
+    f = Forwarder(
+        config.get('api_key', 'fake_api'),
```
Comment (hush-hush) [inline agent.py]: We should fail directly and exit if no api_key is set.

File: agent.py [hunk 4]
```diff
@@ -22,24 +27,52 @@ def init_agent():
...
+        config.get('dd_url', 'https://localhost:9999')
```
Comment (hush-hush) [inline agent.py]: same here.

File: agent.py [hunk 5]
```diff
@@ -22,24 +27,52 @@ def init_agent():
...
+    )
     f.start()
 
-    metadata = get_metadata(get_hostname())
+    # aggregator
+    aggregator = MetricsAggregator(
+        hostname,
+        interval=config.get('histogram_aggregates', 1.0),
```
Comment (hush-hush) [inline agent.py]: Default should be set in the config package like for the other histogram configs

File: collector/collector.py (hunk 2 above)
Comment (truthbk) [reply collector/collector.py]: I'll let this be for now, checks can submit their own service checks for now if need be

File: serialize/serialize.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            self._forwarder.submit_v1_series(
+                metrics, extra_headers)
+        if service_checks:
+            self._forwarder.submit_v1_service_checks(
+                service_checks, extra_headers)
+        if events:
+            self._forwarder.submit_v1_intake(
+                events, extra_headers)
+
+    def submit_metadata(self):
```
Comment (hush-hush) [inline serialize/serialize.py]: Why split setting and sending the metadata ? This add a state to the serializer and I don't think we should mix series and metadata like in agent5. I would remove `add_meta` all together and just have a `send_metadata` method. This would allow us to have a stateless serializer.

File: serialize/serialize.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+        # TODO: do this right
+        if add_meta:
+            metrics['meta'] = self._metadata
+
+        return (json.dumps(metrics),
+                json.dumps(service_checks),
+                json.dumps(events))
+
+    def serialize_and_push(self, add_meta=False):
```
Comment (hush-hush) [inline serialize/serialize.py]: For an other PR: I think we should have a clearer API for the serializer (one method for service_check, one for event and one for metrics). This would be done by having `Serie`, `Event` and `ServiceCheck` class (this will also avoid the very confusing and unclear format https://github.com/DataDog/dd-agent/blob/5.11.1/checks/__init__.py#L905-L926).
We could move the serie split to the `Serie` class.

Also splitting series, event and service_check should be done in the `aggregator` in my opinion.

What do you think ? I don't think this need to be done in this PR.

File: serialize/tests/fixtures/sc_payload.json [hunk 8]
```diff
@@ -0,0 +1 @@
...\nStart time: 1970-01-01 00:00:00\nStop time: \nExit Status: 0", "check": "supervisord.process.status"}, {"status": 2, "tags": ["supervisord_server:server0", "supervisord_process:toxiproxy"], "timestamp": 1487973679.046116, "id": 882, "host_name": "remi-split-payloads", "message": "Current time: 2017-02-24 22:01:18\nProcess name: toxiproxy\nProcess group: tests\nDescription: Not started\nError log file: /var/log/supervisor/toxiproxy-stderr---supervisor-xAR_Ss.log\nStdout log file: /home/vagrant/var/log/toxiproxy.log\nLog file: /home/vagrant/var/log/toxiproxy.log\nState: STOPPED\nStart time: 1970-01-01 00:00:00\nStop time: \nExit Status: 0", "check": "supervisord.process.status"}, {"status": 0, "tags": ["check:supervisord"], "timestamp": 1487973679.046895, "id": 883, "host_name": "remi-split-payloads", "message": null, "check": "datadog.agent.check_status"}, {"status": 0, "tags": ["check:network"], "timestamp": 1487973679.047384, "id": 884, "host_name": "remi-split-payloads", "message": null, "check": "datadog.agent.check_status"}, {"status": 0, "tags": null, "timestamp": 1487973679.047402, "id": 885, "host_name": "remi-split-payloads", "message": null, "check": "datadog.agent.up"}]
```
Comment (hush-hush) [inline serialize/tests/fixtures/sc_payload.json]: Could we add a simpler JSON for testing, those two are so big that it's impossible to see what is tested.

File: serialize/serialize.py (hunk 7 above)
Comment (truthbk) [reply serialize/serialize.py]: I already had this, or something similar, in mind I was working on the events stuff currently, because I think it's currently breaking. If it's fine with you I will address in a separate PR.

File: checks/corechecks/system/tests/test_cpu.py [hunk 9]
```diff
@@ -46,11 +57,19 @@ def test_cpu_first_run(cpu_count, cpu_times, agent_check):
...
 
     c.check({})
-    assert c.get_metrics() == {
-        'system.cpu.system': [{'tags': None, 'type': 'gauge', 'value': 0.2}],
-        'system.cpu.user': [{'tags': None, 'type': 'gauge', 'value': 0.12}],
-        'system.cpu.wait': [{'tags': None, 'type': 'gauge', 'value': 0.0}],
-        'system.cpu.idle': [{'tags': None, 'type': 'gauge', 'value': 4.2300}],
-        'system.cpu.stolen': [{'tags': None, 'type': 'gauge', 'value': 0.0}],
-        'system.cpu.guest': [{'tags': None, 'type': 'gauge', 'value': 0.0}],
+    metrics = c.aggregator.flush()
```
[the remark this answers is not in the document]
Comment (truthbk) [reply checks/corechecks/system/tests/test_cpu.py]: I agree that we can test what goes into the aggregator, that's why my goal is to include the mock aggregator in integrations-core and use the toolkit. I don't think there's any strong reason to mock the AgentCheck in that scenario.

File: forwarder/forwarder.py [hunk 10]
```diff
@@ -11,20 +11,23 @@
...
+    V1_SERIES_ENDPOINT = "/api/v1/series"
+    V1_SERVICE_CHECKS_ENDPOINT = "/api/v1/check_run"
 
     DD_API_HEADER = "DD-Api-Key"
 
     QUEUES_SIZE = 100
     WORKER_JOIN_TIME = 2
 
-    def __init__(self, api_key, domain, nb_worker=4):
+    def __init__(self, api_key, domains, nb_worker=4):
```
[the remark this answers is not in the document]
Comment (truthbk) [reply forwarder/forwarder.py]: I think we should end up implementing it, but I can remove it.

File: serialize/tests/fixtures/sc_payload.json (hunk 8 above)
Comment (truthbk) [reply serialize/tests/fixtures/sc_payload.json]: I pretty-printed the JSON fixture, I think it's now quite clear what's getting tested, especially if you take a look at the test code.

File: serialize/serialize.py (hunk 6 above)
Comment (hush-hush) [reply serialize/serialize.py]: I think that in the end metadata and series will have a different flow in the agent. The metadata don't need aggregation and will most likely not be collect by the `collector`. So the serializer will receive payload from different part of the code at different time. Also series and metadata require very difference processing: they don't use the same endpoint, might not use the same format and `series` require more post-processing.

In the end I think they were merged together in agent5 because we had one endpoint being able to process everything.

File: serialize/serialize.py (hunk 7 above)
Comment (hush-hush) [reply serialize/serialize.py]: Yes, let's open a card for later :+1:

File: checks/corechecks/system/tests/test_cpu.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hush-hush) [reply checks/corechecks/system/tests/test_cpu.py]: On one side I don't expect much more internal checks for the AIX agent (we could even at some point move the one here to integration-core) so I'm fine with that. On the other side if we keep them here this force us to add a dependencies for very simple tests. That's why I found the AgentCheck mock a good in-between.

I guess for now it doesn't really matter: let's merge !
